### PR TITLE
Raise Orchestrator errors on redis errors in orchestrator

### PR DESCRIPTION
**Pull request by kleesc**
### Description of Changes

* Re-raise the exceptions when using the redis orchestrator so that the build manager can handle them.
* Add missing `yield From` on some async calls
* Update the build phase on log failure

#### Changes:

* ..
* ..

#### Issue: https://issues.redhat.com/browse/PROJQUAY-752

**TESTING** ->

**BREAKING CHANGE** ->

---

## Reviewer Checklist

- [ ] It works!
- [ ] Comments provide sufficient explanations for the next contributor
- [ ] Tests cover changes and corner cases
- [ ] Follows Quay syntax patterns and format

### CODE REVIEW

File: buildman/orchestrator.py [hunk 1]
```diff
@@ -784,10 +808,17 @@ def _publish(self, **kwargs):
...
-        yield From(self._client.delete(slash_join(key, REDIS_EXPIRING_SUFFIX)))
-        yield From(self._publish(event=KeyEvent.DELETE, key=key, value=value))
+        try:
+            value = yield From(self._client.get(key))
+            if value is None:
+                raise KeyError
+            yield From(self._client.delete(key))
+            yield From(self._client.delete(slash_join(key, REDIS_EXPIRING_SUFFIX)))
+            yield From(self._publish(event=KeyEvent.DELETE, key=key, value=value))
+        except RedisError as rce:
```
Comment (josephschorr) [inline buildman/orchestrator.py:818]: RedisError handler is repeated here

File: buildman/manager/ephemeral.py [hunk 2]
```diff
@@ -566,6 +566,12 @@ def schedule(self, build_job):
                 realm,
                 build_uuid,
             )
+        except OrchestratorConnectionError:
+            logger.exception(
```
Comment (josephschorr) [inline buildman/manager/ephemeral.py:570]: Does this not need to return or re-raise?

Comment (kleesc) [reply buildman/manager/ephemeral.py:570]: The above one was not being re-raised or returned, so I'm just logging them, since these are just for job metrics.

**Comment by kleesc** [comment]
@josephschorr Ping. `yield From` was missing from some async calls, so it was essentially doing nothing on exceptions... 🤦

File: buildman/orchestrator.py [hunk 3]
```diff
@@ -784,10 +808,17 @@ def _publish(self, **kwargs):
...
         assert not self.is_canceller_only
 
-        value = yield From(self._client.get(key))
-        yield From(self._client.delete(key))
-        yield From(self._client.delete(slash_join(key, REDIS_EXPIRING_SUFFIX)))
-        yield From(self._publish(event=KeyEvent.DELETE, key=key, value=value))
+        try:
+            value = yield From(self._client.get(key))
+            if value is None:
+                raise KeyError
```
Comment (josephschorr) [inline buildman/orchestrator.py:814]: Why this change?

**Comment by kleesc** [comment]
@jzelinskie ping

File: buildman/orchestrator.py (hunk 3 above)
Comment (kleesc) [reply buildman/orchestrator.py:814]: From the interface:
>    Missing keys should return KeyError; otherwise, errors should raise an
    OrchestratorError.

File: buildman/component/buildcomponent.py [hunk 4]
```diff
@@ -360,19 +362,26 @@ def _on_log_message(self, phase, json_data):
                 # If the json data contains an error, then something went wrong with a push or pull.
                 if "error" in log_data:
                     yield From(self._build_status.set_error(log_data["error"]))
+        except RedisError as re:
+            # Move forward with the build if it is a Redis error
+            logger.warning("Failed to set phase on log for build ID: %s - %s", build_id, bse)
+            build_model.update_phase_then_close(build_id, phase)
```
[the remark this answers is not in the document]
Comment (kleesc) [reply buildman/component/buildcomponent.py:368]: @jzelinskie Sure, but it was already yielding control of the event loop to that same synchronous call whenever the worker logged a phase change in `buildstatus.py`.

Since that database call is coupled with the Redis call in `buildstatus.py`, if Redis is down, the build phase won't get updated correctly. 

In this case, we're making sure that we allow the build to move forward, or marked as failure if Redis is not available.

[the remark this answers is not in the document]
Comment (jzelinskie) [reply buildman/component/buildcomponent.py:368]: Ideally, we'd wrap this with the same async wrapper we use for etcd, but I don't think it's worth the effort (and complexity) if we can get by with this until Python 3 or a new build system.
